Feat/dynamic form/site

mvergez opened the pull request.

andriacap committed:
refactor: object.service with observable obj

Change the objectType string to objectType obj
in order to subscribe to multilple properties link to this object

Reviewed-by: andriacap
[Refs_ticket]: #40

andriacap committed:
feat: adapt service to get config json

Get fieldsName and fieldsLabel for display properties
(properties component)

Add logic to setItem in LocalStorage to keepLast Value of observable on reload page

TODO:
- [ ]  see if Localstorage is really necessary with the configService used inside api-geom.service )
- [ ] adapt monitoring-form-g.component with the new service

Reviewed-by: andriacap
[Refs ticket]: #4

andriacap committed:
feat: column datatable from config json

Add logic into sitegroups service in order to use config json to diplay
column datatable (called "display_list")

Review-by: andriac
[Refs ticket]: #4

andriacap committed:
feat: adjust backend to load config json

Adjust backend code to use existing code in order to load config from
file json and by starting with "sites_group"

Fix Media load and upload for site_group

TODO:
- [ ] check if config should be find from file or backend
- [ ] Optimize logic backend (use generic model with class method ?)

Reviewed-by: andriacap
[Refs ticket]: #4

andriacap committed:
feat: add button multiselect with filter on backend

filter with params backend (new route -> see if it's possible to change
that)

Add button multiselect above form type site

TODO:
- improve Input / condition of use case of btn multiselect

Reviewed-by: andriac

andriacap committed:
feat: btn multiselect option

-Add @Input fn , placeholder, title, paramsToFilt
-Remove empty option
-prevent adding from keyboard input or not including in list
-Store config json into object

Reviewed-by: @andriac
[Refs-ticket]: #4

andriacap committed:
refactor: change form-service and object-service

refactor form-service
add observable into object-service (WIP: futur use for refresh page ?)

Rieviewed-by: andriac

[Refs_ticket]: #40

andriacap committed:
refactor: test refresh page and comportment obs

refresh page seems to works with localstorage of different obj in
object_service

Reviewed-by: andriac

andriacap committed:
feat: dynamic form

- Add:
- pass config json to form.component-g.ts
- add config json to this.obj and refresh form into form.component-g.ts
- add css for form component to deal when long form selected

-fix :
- refresh page form component (this._configService.init is necessary
  ...)
- comportment different between add or edit into form component and
  form service

Reviewed-by: andriac
[Refs_ticket]: #4

mvergez commented on backend/gn_module_monitoring/routes/site.py:
```diff
@@ -33,6 +33,26 @@ def get_types_site():
         page=page,
     )
 
+@blueprint.route("/sites/types_label", methods=["GET"])
```
Peut-être plutôt /sites/types/label ? A voir

mvergez commented on backend/gn_module_monitoring/routes/sites_groups.py:
```diff
@@ -66,18 +67,27 @@ def get_sites_group_geometries():
+
+    # db.session.commit()
+    # return item_schema.dump(item), 201
+    # ###############################""
```
Commentaires utiles ?

mvergez commented on backend/gn_module_monitoring/routes/sites_groups.py:
```diff
@@ -91,16 +101,63 @@ def delete(_id):
+    # item = item_schema.load(item_json)
+    # db.session.add(item)
+    # db.session.commit()
+    # return item_schema.dump(item), 201
```
Supprimer commentaires s'ils ne sont plus utilisés

mvergez commented on config/monitoring/generic/site_bak.json:
Fichier utile ?

mvergez commented on frontend/app/class/btn-multiselect-chip.ts:
```diff
@@ -0,0 +1,31 @@
+
+const LIMIT = 50;
+
+type callbackFunction = (pageNumber: number, limit:number, filters: JsonData) => Observable<IPaginated<ISitesGroup> | IPaginated<ISite>>;
```
Le mettre plutôt dans les types non ?

andriacap replied:
Du coup on regroupe tous les "types"  function dans le dossier `gn_module_monitoring/frontend/app/types ` . 

Comme on a ici également : 
https://github.com/NaturalSolutions/gn_module_monitoring/blob/f4220c01a16d1e96b3a44c54b6afaab1b1caf6ff/frontend/app/class/monitoring-geom-component.ts#L4-L6

mvergez replied:
Effectivement, c'est juste utilisé localement, OK

mvergez commented on frontend/app/class/monitoring-site.ts:
```diff
@@ -1,3 +1,4 @@
+// TODO: voir si colonnes doivent être définies depuis le fichier de config site_group
 export enum columnNameSite {
   base_site_name = "Nom",
   last_visit = "Dernière visite",
```
A voir mais sûrement qu'il faut que ce soit dans le fichier de config site_group pour personnaliser les colonnes. Si pas besoin de personnaliser, on peut le laisser là

andriacap replied:
Je me suis justement basé sur le fichier site.json du dossier generic. Toute la logique se base sur ça maintenant

mvergez commented on frontend/app/components/monitoring-datatable-g/monitoring-datatable-g.component.ts:
```diff
@@ -198,6 +199,9 @@ export class MonitoringDatatableGComponent implements OnInit {
   navigateToDetail(row) {
+    // this._objService.changeObjectTypeParent(
+    //   this.objectType
+    // );
```
Commentaire à retirer

mvergez commented on frontend/app/components/monitoring-form-g/monitoring-form.component-g.ts:
```diff
@@ -76,6 +74,7 @@ export class MonitoringFormComponentG implements OnInit {
           // return this._route.queryParamMap;
```
Peut-être utiliser rxjs pour éviter les subscribes à la chaine ?

mvergez commented on frontend/app/components/monitoring-properties-g/monitoring-properties-g.component.ts:
```diff
@@ -1,49 +1,83 @@
   dataDetails: ISitesGroup;
+  fields: JsonData;
+  fieldDefinitions: JsonData;
+  fieldsNames: [];
```
On sait de quelle type de liste c'est ? String ?

mvergez commented on frontend/app/components/monitoring-properties-g/monitoring-properties-g.component.ts:
```diff
@@ -1,49 +1,83 @@
+    //   });
+    // this._objService.currentObjSelected.subscribe((objSelect)=> console.log(objSelect))
     this._objService.currentObjectTypeParent.subscribe((newObjType) => {
+      console.log(newObjType);
```
A enlever (console.log et commentaire)

mvergez commented on frontend/app/components/monitoring-properties-g/monitoring-properties-g.component.ts:
```diff
@@ -1,49 +1,83 @@
+      this.fields = newObjType.template.fieldLabels;
+      this.fieldDefinitions = newObjType.template.fieldDefinitions;
+      this.objectType.properties = this.selectedObj;
+      console.log(this.objectType);
```
idem, à enlever aussi

mvergez commented on frontend/app/gnModule.module.ts:
```diff
@@ -79,12 +83,15 @@ const routes: Routes = [
         children: [
           {
-            path: "",
+            path:"",
```
Détail, mais normalement il faudrait grader l'espace non ?

mvergez commented on frontend/app/interfaces/geom.ts:
```diff
@@ -44,7 +44,13 @@ export interface IGeomService {
-  create(postdata: IGeomObject): Observable<Resp>;
-  patch(id: number, updatedData: IGeomObject): Observable<Resp>;
-  // delete(obj: IGeomObject)
+  create(postdata: { properties: IGeomObject }): Observable<Resp>;
```
Il faudrait créer une interface plutôt. A voir si c'est valable pour tous les objets de Monitoring

mvergez commented on frontend/app/components/btn-select/btn-select.component.ts:
```diff
@@ -0,0 +1,136 @@
+      debounceTime(400),
+      distinctUntilChanged(),
+      switchMap((val: string) => {
+        console.log(val);
```
console.log à retirer

mvergez commented on frontend/app/components/btn-select/btn-select.component.ts:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  remove(option: string): void {
+    console.log(this.configObjAdded[option]);
```
console.log à retirer

mvergez commented on frontend/app/components/btn-select/btn-select.component.ts:
```diff
@@ -0,0 +1,136 @@
+
+    if (this.configObjAdded && this.configObjAdded[option] !== undefined) {
+      delete this.configObjAdded[option];
+      console.log(this.configObjAdded);
```
console.log à retirer

mvergez commented on frontend/app/components/btn-select/btn-select.component.ts:
```diff
@@ -0,0 +1,136 @@
+      ? this.listOptionChosen.push(event.option.viewValue) &&
+        this.addObject(event.option.value)
+      : null;
+    console.log(this.listOptionChosen);
```
console.log à retirer

mvergez commented on frontend/app/components/btn-select/btn-select.component.ts:
```diff
@@ -0,0 +1,136 @@
+      // Ici on map pour uniformiser la "key" utilisé pour afficher les options (default Key : 'name')
+      map((response) =>
+        response.filter((obj) => {
+          console.log(obj);
```
console.log à retirer

mvergez commented on frontend/app/interfaces/objObs.ts:
```diff
@@ -0,0 +1,24 @@
+export interface IobjObs<ObjDataType> {
+  properties: ObjDataType;
+  endPoint: endPoints;
+  objectType: "site" | "sites_group";
```
Ce ne sera pas compatible avec les visites, c'est pas grave ?
objectType et objDataType n'est pas redondant ?

mvergez replied:
A voir avec les prochaines PR où il y a un refact de cela normalement

mvergez commented on frontend/app/services/config-json.service.ts:
Ce n'est pas redondant avec config.service.ts ? Il ne faudrait pas le modifier pour l'adapter ?

andriacap replied:
Yes dans l'idéal ça serait ça mais le service est très dépendant  à l'objet Monitoring . A voir ce qu'on peut faire

mvergez replied:
OK, j'ai mis un commentaire FIXME pour voir si c'est mergeable avec le `config.service.ts`

mvergez commented on frontend/app/services/config.service.ts:
```diff
@@ -127,7 +127,7 @@ export class ConfigService {
   schema(moduleCode, objectType, typeSchema = 'all'): Object {
     moduleCode = moduleCode || 'generic';
-
+    console.log(objectType)
```
A retirer

mvergez commented on frontend/app/services/data-table.service.ts:
```diff
@@ -39,8 +39,9 @@ export class DataTableService {
-  colsTable(colName:IColumn[], dataTable): IColumn[] {
+  colsTable(colName:{}, dataTable): IColumn[] {
     const arr = Object.keys(colName);
+    console.log(arr)
```
A retirer.
colName ne peut pas avoir de type ?

mvergez commented on frontend/app/services/object.service.ts:
```diff
@@ -1,23 +1,68 @@
+    console.log("Change selected Obj",newObjSelected)
+    if (storeObjectTypeSelected)
+      localStorage.setItem('storedObjectSelected', JSON.stringify(newObjSelected));
+     this.objSelected.next(newObjSelected);
```
A voir si on ne peut pas utiliser le `cache.service.ts` qui implémente des fonctions de caching déjà

mvergez commented on backend/gn_module_monitoring/monitoring/definitions.py:
```diff
@@ -41,3 +41,26 @@
+    'sites_group': MonitoringObjectGeom,
+}
+
+monitoring_g_definitions.set(MonitoringObjectsG_dict, MonitoringModelsG_dict)
```
Il faudrait peut être plus s'inspirer des dict d'au dessus comme ils sont quasi identiques. Et voir s'il est vraiment nécessaire de créer une autre définitions

andriacap committed:
feat: dynamic form - Correction PR

Remove unused console.log
Rxjs : use concatMap to avoid subscribe inside subscribe
Apply: prettier and sort prettier for import ts file

Reviewed-by: @andriac
[Refs_PR]: #42

andriacap committed:
feat: dynamic create site

- Add change current-object when click "Add <object>" from datatable
  component
- api-geom.service: change the way to init the api-geom
- WIP:  two subscribes in one .. (the subscription is call only once ..)
- WIP : get config json from backend

Reviewed-by: andriac

andriacap committed:
feat: add relationship id into sendData form site

Add type_site and id_site_group to the form site

Reviewed-by: andriac
[Refs_ticket]: #5 et #6

andriacap committed:
feat(back): add custom config when post site

Change backend files in order to choose specific config when site is
created.

Reviewed-by: andriac
[Refs_ticket]: #5 , #6

andriacap committed:
fix: problem of type object when loading form

Add endPoint and objectType to the observable using by the formservice
in order to use _apiGeomService with good context (endPoint)

[Refs_ticket]: #5 et #6

andriacap committed:
fix: todo comments about refactoring code

Add todo comment in order to don't forget what
is to improve inside the code

Commit:
chore(front): add comment and remove console.log

Commit:
chore(api): put back old formatting

Commit:
chore(api): remove useless comment

Commit:
refactor(api): move and rename create_update fct

Commit:
fix: put back inventor/digitiser & fix typo

Commit:
chore: remove useless files

Since columns are now loaded from json, they are not needed
anymore in separate typescript files

Commit:
chore: remove definition from sites_group.json

Commit:
fix: column problem on sites

Commit:
chore: remove useless comment

Commit:
refactor(front): remove comment and use rxjs

Commit:
chore(front): remove useless file

Since class BtnMultiSelectChipClass is not used anymore

Commit:
chore(front): remove useless comments/imports

Commit:
style(front): applied prettier

Commit:
chore(front): remove useless code

Commit:
refactor(front): regroup types into an interface

Commit:
style(front): applied prettier

Commit:
refactor(front): inherit config-json from config

andriacap reviewed (approved):
Vérifié avec Maxime
